"""
Vue principale de l'application PyTasks.
Connecte l'interface utilisateur au contr√¥leur.
"""
from PySide6.QtWidgets import QMainWindow, QListWidgetItem, QInputDialog, QMessageBox
from PySide6.QtCore import Qt, Slot
from datetime import datetime

from views.ui_main import Ui_MainWindow
from controllers.task_controller import TaskController
from models.task import Task, TaskState


class MainWindow(QMainWindow):
    """
    Fen√™tre principale de l'application.
    G√®re l'affichage et les interactions utilisateur.
    """
    
    def __init__(self, controller: TaskController):
        print("  üî® MainWindow.__init__() appel√©")
        
        try:
            super().__init__()
            print("  ‚úÖ super().__init__() OK")
        except Exception as e:
            print(f"  ‚ùå Erreur super().__init__(): {e}")
            raise
        
        try:
            # Setup UI
            self.ui = Ui_MainWindow()
            print("  ‚úÖ Ui_MainWindow() cr√©√©")
        except Exception as e:
            print(f"  ‚ùå Erreur cr√©ation Ui_MainWindow: {e}")
            raise
        
        try:
            self.ui.setupUi(self)
            print("  ‚úÖ setupUi() termin√©")
        except Exception as e:
            print(f"  ‚ùå Erreur setupUi(): {e}")
            import traceback
            traceback.print_exc()
            raise
        
        try:
            # Contr√¥leur
            self.controller = controller
            print("  ‚úÖ Contr√¥leur assign√©")
            
            # √âtat interne
            self.current_state_filter = None
            print("  ‚úÖ √âtat interne initialis√©")
            
            # Connecter les signaux
            print("  üîå Connection des signaux...")
            self._connect_signals()
            print("  ‚úÖ Signaux connect√©s")
            
            # Charger les donn√©es initiales
            print("  üìÇ Chargement des donn√©es...")
            self.controller.load_tasks()
            print("  ‚úÖ Donn√©es charg√©es")
            
        except Exception as e:
            print(f"  ‚ùå Erreur dans l'initialisation: {e}")
            import traceback
            traceback.print_exc()
            raise
    
    # ========== CONNEXION DES SIGNAUX ==========
    
    def _connect_signals(self):
        """Connecte tous les signaux de l'UI au contr√¥leur"""
        
        # === SIDEBAR ===
        # Recherche
        self.ui.searchBar.textChanged.connect(self._on_search_changed)
        
        # Filtre par √©tat
        self.ui.stateFilter.currentTextChanged.connect(self._on_filter_changed)
        
        # S√©lection d'une t√¢che
        self.ui.taskList.itemClicked.connect(self._on_task_selected)
        
        # Boutons CRUD
        self.ui.btnAdd.clicked.connect(self._on_add_task)
        self.ui.btnDelete.clicked.connect(self._on_delete_task)
        
        # === D√âTAILS T√ÇCHE ===
        # Boutons actions
        self.ui.btnSave.clicked.connect(self._on_save_task)
        self.ui.btnClose.clicked.connect(self._on_close_task)
        
        # Commentaires
        self.ui.btnAddComment.clicked.connect(self._on_add_comment)
        self.ui.commentInput.returnPressed.connect(self._on_add_comment)  # Entr√©e = ajouter
        
        # === HISTORIQUE ===
        self.ui.btnClearHistory.clicked.connect(self._on_clear_history)
        
        # === SIGNAUX DU CONTR√îLEUR ===
        # Quand la liste change
        self.controller.tasks_updated.connect(self._refresh_task_list)
        
        # Quand une t√¢che est s√©lectionn√©e
        self.controller.task_selected.connect(self._display_task_details)
    
    # ========== RECHERCHE & FILTRE ==========
    
    @Slot()
    def _on_search_changed(self):
        """D√©clench√© quand le texte de recherche change"""
        query = self.ui.searchBar.text()
        self.controller.search_and_filter(query, self.current_state_filter)
    
    @Slot(str)
    def _on_filter_changed(self, state_text: str):
        """D√©clench√© quand le filtre d'√©tat change"""
        # Conversion texte -> TaskState
        state_map = {
            "Tous les √©tats": None,
            "√Ä faire": TaskState.TODO,
            "En cours": TaskState.IN_PROGRESS,
            "R√©alis√©": TaskState.DONE,
            "Abandonn√©": TaskState.ABANDONED,
            "En attente": TaskState.WAITING
        }
        
        self.current_state_filter = state_map.get(state_text)
        
        # Applique le filtre
        query = self.ui.searchBar.text()
        self.controller.search_and_filter(query, self.current_state_filter)
    
    # ========== AFFICHAGE LISTE ==========
    
    @Slot()
    def _refresh_task_list(self):
        """Rafra√Æchit la liste des t√¢ches affich√©es"""
        self.ui.taskList.clear()
        
        tasks = self.controller.get_all_tasks()
        
        for task in tasks:
            # Ic√¥ne selon l'√©tat
            icon_map = {
                TaskState.TODO: "üìã",
                TaskState.IN_PROGRESS: "‚öôÔ∏è",
                TaskState.DONE: "‚úÖ",
                TaskState.ABANDONED: "‚ùå",
                TaskState.WAITING: "‚è≥"
            }
            icon = icon_map.get(task.state, "üìã")
            
            # Cr√©e l'item
            item = QListWidgetItem(f"{icon} {task.title}")
            item.setData(Qt.UserRole, task.id)  # Stocke l'ID dans l'item
            
            self.ui.taskList.addItem(item)
        
        # Met √† jour la barre de statut
        self.statusBar().showMessage(f"{len(tasks)} t√¢che(s)")
    
    # ========== S√âLECTION T√ÇCHE ==========
    
    @Slot(QListWidgetItem)
    def _on_task_selected(self, item: QListWidgetItem):
        """D√©clench√© quand on clique sur une t√¢che"""
        task_id = item.data(Qt.UserRole)
        self.controller.select_task(task_id)
        
        # Active le bouton supprimer
        self.ui.btnDelete.setEnabled(True)
    
    @Slot(Task)
    def _display_task_details(self, task: Task):
        """Affiche les d√©tails d'une t√¢che dans le panneau de droite"""
        # Cache le label "Aucune s√©lection"
        self.ui.noSelectionLabel.setVisible(False)
        
        # Affiche le groupe de d√©tails
        self.ui.taskDetailsGroup.setVisible(True)
        
        # Remplit les champs
        self.ui.titleEdit.setText(task.title)
        self.ui.descriptionEdit.setPlainText(task.description)
        
        # √âtat
        state_index_map = {
            TaskState.TODO: 0,
            TaskState.IN_PROGRESS: 1,
            TaskState.DONE: 2,
            TaskState.ABANDONED: 3,
            TaskState.WAITING: 4
        }
        self.ui.stateEdit.setCurrentIndex(state_index_map.get(task.state, 0))
        
        # Dates
        if task.start_date:
            self.ui.startDateEdit.setDateTime(task.start_date)
        else:
            self.ui.startDateEdit.clear()
        
        if task.end_date:
            self.ui.endDateEdit.setDateTime(task.end_date)
        else:
            self.ui.endDateEdit.clear()
        
        # Commentaires
        self._refresh_comments(task)
    
    def _refresh_comments(self, task: Task):
        """Rafra√Æchit la liste des commentaires"""
        self.ui.commentsList.clear()
        
        for comment in task.comments:
            timestamp = comment.created_at.strftime("%d/%m/%Y %H:%M")
            item_text = f"üí¨ [{timestamp}] {comment.content}"
            self.ui.commentsList.addItem(item_text)
    
    # ========== CR√âATION T√ÇCHE ==========
    
    @Slot()
    def _on_add_task(self):
        """D√©clench√© par le bouton Ajouter"""
        # Bo√Æte de dialogue pour le titre
        title, ok = QInputDialog.getText(
            self,
            "Nouvelle t√¢che",
            "Titre de la t√¢che :"
        )
        
        if ok and title:
            # Bo√Æte de dialogue pour la description (optionnel)
            description, ok = QInputDialog.getMultiLineText(
                self,
                "Nouvelle t√¢che",
                "Description (optionnel) :"
            )
            
            if ok:
                success = self.controller.create_task(title, description)
                
                if success:
                    self.statusBar().showMessage("‚úÖ T√¢che cr√©√©e !", 3000)
    
    # ========== MODIFICATION T√ÇCHE ==========
    
    @Slot()
    def _on_save_task(self):
        """Sauvegarde les modifications de la t√¢che actuelle"""
        if not self.controller.current_task:
            return
        
        # R√©cup√®re les valeurs des champs
        title = self.ui.titleEdit.text()
        description = self.ui.descriptionEdit.toPlainText()
        
        # √âtat
        state_map = [
            TaskState.TODO,
            TaskState.IN_PROGRESS,
            TaskState.DONE,
            TaskState.ABANDONED,
            TaskState.WAITING
        ]
        state = state_map[self.ui.stateEdit.currentIndex()]
        
        # Dates (peut √™tre None)
        start_date = self.ui.startDateEdit.dateTime().toPython() if self.ui.startDateEdit.dateTime().isValid() else None
        end_date = self.ui.endDateEdit.dateTime().toPython() if self.ui.endDateEdit.dateTime().isValid() else None
        
        # Appelle le contr√¥leur
        success = self.controller.update_current_task(
            title=title,
            description=description,
            start_date=start_date,
            end_date=end_date,
            state=state
        )
        
        if success:
            self.statusBar().showMessage("üíæ T√¢che enregistr√©e !", 3000)
    
    # ========== SUPPRESSION T√ÇCHE ==========
    
    @Slot()
    def _on_delete_task(self):
        """Supprime la t√¢che s√©lectionn√©e"""
        current_item = self.ui.taskList.currentItem()
        if not current_item:
            return
        
        task_id = current_item.data(Qt.UserRole)
        
        if self.controller.delete_task(task_id):
            # Cache les d√©tails
            self._hide_task_details()
            
            # D√©sactive le bouton supprimer
            self.ui.btnDelete.setEnabled(False)
            
            self.statusBar().showMessage("üóëÔ∏è T√¢che supprim√©e", 3000)
    
    # ========== CL√îTURE T√ÇCHE ==========
    
    @Slot()
    def _on_close_task(self):
        """Cl√¥ture la t√¢che actuelle"""
        if not self.controller.current_task:
            return
        
        # Demande confirmation
        reply = QMessageBox.question(
            self,
            "Cl√¥turer la t√¢che",
            f"Cl√¥turer la t√¢che '{self.controller.current_task.title}' ?\n\n"
            "Elle sera marqu√©e comme 'R√©alis√©' avec la date de fin actuelle.",
            QMessageBox.Yes | QMessageBox.No
        )
        
        if reply == QMessageBox.Yes:
            success = self.controller.close_current_task()
            
            if success:
                # Rafra√Æchit l'affichage
                self.controller.select_task(self.controller.current_task.id)
                self.statusBar().showMessage("‚úÖ T√¢che cl√¥tur√©e !", 3000)
    
    # ========== COMMENTAIRES ==========
    
    @Slot()
    def _on_add_comment(self):
        """Ajoute un commentaire √† la t√¢che actuelle"""
        if not self.controller.current_task:
            return
        
        content = self.ui.commentInput.text().strip()
        
        if not content:
            return
        
        success = self.controller.add_comment_to_current_task(content)
        
        if success:
            # Vide le champ
            self.ui.commentInput.clear()
            
            # Rafra√Æchit les commentaires
            self._refresh_comments(self.controller.current_task)
            
            self.statusBar().showMessage("üí¨ Commentaire ajout√©", 2000)
    
    # ========== HISTORIQUE ==========
    
    @Slot()
    def _on_clear_history(self):
        """Efface l'historique des logs"""
        reply = QMessageBox.question(
            self,
            "Effacer l'historique",
            "Voulez-vous effacer tout l'historique ?",
            QMessageBox.Yes | QMessageBox.No
        )
        
        if reply == QMessageBox.Yes:
            self.controller.logger.clear()
            self.ui.historyLog.clear()
            self.statusBar().showMessage("üóëÔ∏è Historique effac√©", 2000)
    
    def update_history_display(self):
        """Met √† jour l'affichage de l'historique"""
        logs = self.controller.logger.get_all_logs()
        self.ui.historyLog.setPlainText(logs)
    
    # ========== UTILITAIRES ==========
    
    def _hide_task_details(self):
        """Cache le panneau de d√©tails"""
        self.ui.taskDetailsGroup.setVisible(False)
        self.ui.noSelectionLabel.setVisible(True)